app/src/main/java/com/nextcloud/talk/shareditems/adapters/SharedItemsAdapter.kt

```diff
@@ -0,0 +1,146 @@
+package com.nextcloud.talk.shareditems.adapters
+
+import android.graphics.drawable.Drawable
+import android.net.Uri
+import android.text.format.Formatter
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.core.content.ContextCompat
+import androidx.recyclerview.widget.RecyclerView
+import com.facebook.drawee.backends.pipeline.Fresco
+import com.facebook.drawee.controller.BaseControllerListener
+import com.facebook.drawee.controller.ControllerListener
+import com.facebook.drawee.interfaces.DraweeController
+import com.facebook.drawee.view.SimpleDraweeView
+import com.facebook.imagepipeline.common.RotationOptions
+import com.facebook.imagepipeline.image.ImageInfo
+import com.facebook.imagepipeline.request.ImageRequestBuilder
+import com.nextcloud.talk.databinding.SharedItemGridBinding
+import com.nextcloud.talk.databinding.SharedItemListBinding
+import com.nextcloud.talk.shareditems.model.SharedItem
+import com.nextcloud.talk.utils.DateUtils
+import com.nextcloud.talk.utils.DrawableUtils
+import com.nextcloud.talk.utils.FileViewerUtils
+
+class SharedItemsAdapter(private val showGrid: Boolean) : RecyclerView.Adapter<SharedItemsViewHolder>() {
+
+    companion object {
+        private val TAG = SharedItemsAdapter::class.simpleName
+        private const val ONE_SECOND_IN_MILLIS = 1000
+    }
+
+    class ViewHolder(val binding: SharedItemGridBinding, itemView: View) : RecyclerView.ViewHolder(itemView)
+
+    var authHeader: Map<String, String> = emptyMap()
+    var items: List<SharedItem> = emptyList()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SharedItemsViewHolder {
+
+        return if (showGrid) {
+            SharedItemsGridViewHolder(
+                SharedItemGridBinding.inflate(
+                    LayoutInflater.from(parent.context),
+                    parent,
+                    false
+                )
+            )
+        } else {
+            SharedItemsListViewHolder(
+                SharedItemListBinding.inflate(
+                    LayoutInflater.from(parent.context),
+                    parent,
+                    false
+                )
+            )
+        }
+    }
+
+    override fun onBindViewHolder(holder: SharedItemsViewHolder, position: Int) {
+
+        val currentItem = items[position]
+
+        holder.image.hierarchy.setPlaceholderImage(staticImage(currentItem.mimeType, holder.image))
+        if (currentItem.previewAvailable == true) {
+            holder.image.controller = configurePreview(currentItem, holder)
+        }
+
+        when (holder) {
+            is SharedItemsListViewHolder -> {
+                bindListViewHolder(holder, currentItem)
+            }
+        }
```

Comment:
The Adapter should not care what subtype the ViewHolder is after creating them (liskov substitution principle). If you need to do extra things only to the list viewholder, it's better to have an abstract method on the shared viewholder class, and call it here without checking which class it is.

The implementation can be empty by default and only implemented in the list viewholder.

Reply:
I implement your suggestion. But I'm not very happy with the name of the new function `SharedItemsViewHolder#specificBindings`. 

Do you have a better one?

Reply:
I can't think of a clearly better name (maybe `onBind` or something), which usually means the design could be better.

Here's an alternate idea: why not move the entirety of the `onBindViewHolder` (with the item as a parameter) to the base ViewHolder? that way the list viewholder can override it, call the super method, and add the extra logic at the end.

Reply:
> Here's an alternate idea: why not move the entirety of the `onBindViewHolder` (with the item as a parameter) to the base ViewHolder? that way the list viewholder can override it, call the super method, and add the extra logic at the end.

I did it that way. But now the `userName` and `userToken` must be passed via the `SharedItemsViewHolder` to create the HTTP headers.